### PR TITLE
[Clarissa Lee Zi Xin] iP

### PR DESCRIPTION
**Pull request by clarlzx**
**What is the purpose of this pull request?**

- Chat the Cat replaces Duke as the personality behind the chat bot
- Additional functionalities are added to Chat the Cat

### CODE REVIEW

File: src/main/java/chat/Parser.java [hunk 1]
```diff
@@ -1,11 +1,6 @@
 package chat;
 
-import chat.command.Command; 
-import chat.command.ListCommand; 
-import chat.command.DoneCommand;
-import chat.command.DeleteCommand;
-import chat.command.AddCommand;
-import chat.command.ExitCommand;
+import chat.command.*;
```
Comment (JodyLorah) [inline src/main/java/chat/Parser.java]: Maybe list out all imports explicitly instead of *

File: src/main/java/chat/command/FindCommand.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package chat.command;
+
+import chat.*;
+import chat.task.Task;
+
+import java.util.ArrayList;
+
+public class FindCommand extends Command {
```
Comment (JodyLorah) [inline src/main/java/chat/command/FindCommand.java:15]: Maybe add some javadocs so others would know what youre trying to do in this class

File: src/main/java/chat/Storage.java [hunk 3]
```diff
@@ -56,7 +56,7 @@ public void save(TaskList tasks) throws ChatException {
      */
     public ArrayList<Task> load() throws ChatException {
         try {
-            ArrayList<Task> taskListFromFile = new ArrayList<>();
+            ArrayList<Task> tasksFromFile = new ArrayList<>();
```
Comment (JodyLorah) [inline src/main/java/chat/Storage.java]: Nice use of plural since it is an arraylist of tasks

File: src/main/java/chat/task/Task.java [hunk 4]
```diff
@@ -30,20 +30,20 @@ public String getName() {
...
-     * Returns boolean done that shows if task has been completed.
+     * Returns boolean isDone that shows if task has been completed.
      * If true, then task is complete.
      * 
-     * @return Boolean done that tells if task is completed.
+     * @return Boolean isDone that tells if task is completed.
      */
-    public boolean getDone() {
-        return this.done;
+    public boolean getIsDone() {
```
Comment (JodyLorah) [inline src/main/java/chat/task/Task.java:40]: Maybe name the boolean as a question isTaskDone

File: src/main/java/chat/task/Todo.java [hunk 5]
```diff
@@ -62,13 +62,13 @@ public static Todo createTodo (String str) throws ChatException {
...
-        return String.format("T,%s,%s", this.getDone(), this.getName());
+        return String.format("T,%s,%s", this.getIsDone(), this.getName());
     }
 
     /**
      * Returns a string that shows the details of the task.
-     * <p>[ ] will be displayed if done = false.</p>
-     * <p>[X] will be displayed if done = true.</p>
+     * <p>[ ] will be displayed if isDone = false.</p>
+     * <p>[X] will be displayed if isDone = true.</p>
```
Comment (JodyLorah) [inline src/main/java/chat/task/Todo.java:103]: I feel dont need to so explicitly show what will be returned since the toString method is quite clear

**Review by JodyLorah — COMMENTED** [review]
Generally looks good and neat, but javadocs lacking in some classes,

File: src/main/java/chat/Ui.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    private static String goCat = "(=^. .^=)~~";
+    private static String gdJobCat = "\\(=^> <^=)/";
+    private static String errorCat = "(=^> <^=)'''";
+    
+    private static Scanner sc = new Scanner(System.in);
+
+    /**
+     * Prints lines that divide each response.
+     */
+    public void lines() {
```
Comment (markmcwong) [inline src/main/java/chat/Ui.java]: Perhaps naming the method as printLine() since it is not operating on a collection of objects (not lines but rather only one line) and use verb at the start also?

File: src/main/java/chat/Parser.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            return new DoneCommand(str);
+        } else if (str.startsWith("todo") | str.startsWith("deadline") | str.startsWith("event")) {
+            return new AddCommand(str);
+        } else if (str.startsWith("delete")) {
+            return new DeleteCommand(str);
+        } else if (str.equals("bye")) {
+            return new ExitCommand();
+        } else if (str.startsWith("find")) {
+            return new FindCommand(str);
+        } else {
```
Comment (markmcwong) [inline src/main/java/chat/Parser.java:39]: Perhaps using a switch statement will be clear and easier to maintain?

File: src/main/java/chat/command/FindCommand.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
+package chat.command;
+
+import chat.*;
```
Comment (markmcwong) [inline src/main/java/chat/command/FindCommand.java]: Perhaps avoid using the wild imports and list imported classes explicitly?

File: src/main/java/chat/task/Deadline.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import java.time.format.DateTimeParseException;
+
+import chat.ChatException;
+
+/**
+ * Deadline is a type of <b>task</b> with an <b>end</b> date and time.
+ */
+public class Deadline extends Task {
+
+    private LocalDateTime end;
```
Comment (markmcwong) [inline src/main/java/chat/task/Deadline.java:14]: Perhaps naming "end" as a more descriptive name such endDate ?

File: src/main/java/chat/task/Deadline.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            throw new ChatException("deadline name missing\n" + formatStr);
+        } else if (!tempStr.contains("/by")) {
+            //i.e. deadline return book
+            //i.e. deadline return book tmr
+            throw new ChatException("/by missing\n" + formatStr);
+        } else if (!tempStr.contains(" /by ")) {
+            //i.e. deadline return book/by
+            //i.e. deadline return book/by tmr
+            //i.e. deadline return book /bytmr
+            //i.e. deadline return book/bytmr
```
Comment (markmcwong) [inline src/main/java/chat/task/Deadline.java:125]: nice description of what the `if` condition will filter out with examples

**Review by markmcwong — COMMENTED** [review]
clear and well documented, with examples in comments to make readers easier to understand

File: src/main/java/chat/Chat.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+            ui.showLoadingError();
+            tasks = new TaskList();
+        }
+    }
+
+    /**
+     * Method that starts and runs Chat the Cat.
+     */
+    public void run() {
+        ui.lines();
```
Comment (nicholasnge) [inline src/main/java/chat/Chat.java]: Could consider calling lines() in the greet() method (and other places where you call it) instead of in the main, where lines are very precious. ie. in the greet method:
`greet() {
ui.lines();
...
ui.lines()
}`

File: src/main/java/chat/Chat.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+import chat.command.Command;
+import chat.TaskList;
+import chat.Storage;
+import chat.Ui;
+
+/**
+ * Main class that executes Chat the Cat.
+ */
+
+public class Chat {
```
Comment (nicholasnge) [inline src/main/java/chat/Chat.java:11]: love the concept of Chat the cat

File: src/main/java/chat/Ui.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+import java.util.Scanner;
+
+import chat.task.Task;
+
+/**
+ * Ui is a class that deals with interactions with the user.
+ */
+public class Ui {
+    
+    private static String greetingCat = "(=^. .^=)";
```
Comment (nicholasnge) [inline src/main/java/chat/Ui.java]: incredible stuff

File: src/main/java/chat/Ui.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+    /**
+     * Displays tasks from list of tasks from TaskList object.
+     * 
+     * @param tasks TaskList object that contains a list of tasks.
+     */
+    public void list(TaskList tasks) {
+        int i = 0;
+        System.out.println(" * list *");
+        while (i < tasks.getTasks().size()) {
```
Comment (nicholasnge) [inline src/main/java/chat/Ui.java]: i think its more standard to use the standard for loop in this scenario, even though they do the same thing... so that the code can be less surprising and more standard

File: src/main/java/chat/command/Command.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+     * 
+     * @param tasks TaskList object that contains a list of tasks.
+     * @param commandName Command name.
+     * @param str Inputted command string from user to Chat the Cat.
+     * @return Task.
+     * @throws ChatException If format of command is wrong.
+     */
+    public Task checkCommandIndex(TaskList tasks, String commandName, String str) throws ChatException {
+        //for commands with format: [command] [index]
+        //returns valid index if index is correct
```
Comment (nicholasnge) [inline src/main/java/chat/command/Command.java:45]: the structure here is abit hard to follow... since _"Please input with this ... "_ occurs quite alot in the exceptions, maybe can refactor that phrase out of all the exceptions which point to formatting error?

File: src/main/java/chat/task/Deadline.java [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+     */
+    public String allParameterStr() {
+        return String.format("D,%s,%s,%s", this.getIsDone(), this.getName(), this.getEnd().format(inputFormatter));
+    }
+
+    /**
+     * Returns a string that shows the details of the task.
+     * <p>[ ] will be displayed if isDone = false.</p>
+     * <p>[X] will be displayed if isDone = true.</p>
+     *
```
Comment (nicholasnge) [inline src/main/java/chat/task/Deadline.java:176]: seriously very nice docs

File: src/test/java/chat/task/EventTest.java [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+        try {
+            Event.createEvent("event       ");
+            fail();
+        } catch (ChatException e) {
+            assertEquals(errorMessage, e.getMessage());
+        }
+    }
+
+    @Test
+    public void createEvent_noSpaceBetweenEventAndDescription_chatException() {
```
Comment (nicholasnge) [inline src/test/java/chat/task/EventTest.java:65]: hmm isnt the standard to capitalize the first words, eg. CreateEvent_NoSpace..._ChatException

**Review by nicholasnge — COMMENTED** [review]
very nice comments and examples in the docs

File: src/main/java/chat/Parser.java (hunk 1 above)
Comment (clarlzx) [reply src/main/java/chat/Parser.java]: Yes that seems better, thank you for pointing that out.

File: src/main/java/chat/command/FindCommand.java (hunk 2 above)
Comment (clarlzx) [reply src/main/java/chat/command/FindCommand.java:15]: Oh didn't realize that I missed that out, thank you for pointing it out.

File: src/main/java/chat/Parser.java (hunk 7 above)
Comment (clarlzx) [reply src/main/java/chat/Parser.java:39]: Yep I think that will be better, thanks for the suggestion 😄.

File: src/main/java/chat/command/FindCommand.java (hunk 8 above)
Comment (clarlzx) [reply src/main/java/chat/command/FindCommand.java]: Yes that seems better, thank you for pointing that out.

File: src/main/java/chat/Chat.java (hunk 12 above)
Comment (clarlzx) [reply src/main/java/chat/Chat.java:11]: Hahaha, thanks 😸

File: src/test/java/chat/task/EventTest.java (hunk 17 above)
Comment (clarlzx) [reply src/test/java/chat/task/EventTest.java:65]: Maybe you have remembered wrongly because I found this on the coding standard guide?

image

File: src/main/java/chat/command/Command.java (hunk 15 above)
Comment (clarlzx) [reply src/main/java/chat/command/Command.java:45]: Yes I think that this method could be made cleaner, thanks for the suggestion.
